README
Copyright (C) 2012, Technical University of Denmark

--------------------------------------------------------------------------------

    This file is part of ssparse.

    ssparse is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
    
    ssparse is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
    
    You should have received a copy of the GNU Lesser General Public
    License along with this Module; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 
--------------------------------------------------------------------------------


# ssparse

Small utility library for sparse linear algebra. 

https://github.com/nefan/ssparse

Please note that ssparse is partly derived from CSparse and CXSparse

CXSparse: a Concise Sparse matrix package - Extended.
http://www.suitesparse.com

See also the AUTHORS file and the header of each source file for copyright
information.


## Usage

See e.g. https://github.com/nefan/networkred for examples.


## License

Everything here is LGPL.


## Installation

Confer the INSTALL file.


## Documentation

See e.g. https://github.com/nefan/networkred for examples.


## News

November 2012: project started.


## Credits

This work was created as a part of the Secure Operation of Sustainable Power
Systems (SOSPO) project with support from the Danish Strategic Research Council
(DSF).



HISTORY
Copyright (C) 2012, Technical University of Denmark
https://github.com/nefan/ssparse

--------------------------------------------------------------------------------

    This file is part of ssparse.

    ssparse is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
    
    ssparse is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
    
    You should have received a copy of the GNU Lesser General Public
    License along with this Module; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 
--------------------------------------------------------------------------------

November 2012:
Project started.



src/ssp_spdot.c
/*
 * This file is part of ssparse.
 *
 * Copyright (c) 2006, Timothy A. Davis.
 * File derived from:
 * CXSparse: a Concise Sparse matrix package - Extended.
 * http://www.suitesparse.com
 * 
 * Modified by Stefan Sommer (shso@elektro.dtu.dk), November 2012
 *
 * ssparse is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * ssparse is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this Module; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */
 
#include "ssparse.h"

/* dot product between dense vectors x,y with x having non-zero pattern xi */
CS_ENTRY ssp_spdot (CS_INT *xi, CS_ENTRY *x, CS_ENTRY *y, CS_INT n)
{
    CS_ENTRY v = 0;
    for (int i=0; i<n; i++) {
        int j = xi[i];

        if (y[j] != 0)
            v += CS_CONJ(x[j])*y[j];
    }

    return v;
}


src/ssp_cuda.cpp
/*
 * This file is part of ssparse.
 *
 * Copyright (C) 2012, Technical University of Denmark
 * https://github.com/nefan/ssparse
 *
 * ssparse is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * ssparse is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this Module; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "ssparse.h"

// initialize CUDA
ssp_cuda *ssp_init_cuda() {
    ssp_cuda *cudaHandle = (ssp_cuda*)malloc(sizeof(ssp_cuda));
    if (!cudaHandle) {
        fprintf(stderr,"ssp_init_cuda: cudaHandle memory allocation failed.\n");
        return NULL;
    }
    cudaHandle->cusparse_handle = 0;
    cudaHandle->cusparse_matDescr = 0;

    cusparseStatus_t status = cusparseCreate(&cudaHandle->cusparse_handle);

    if (status != CUSPARSE_STATUS_SUCCESS) {
        ssp_finalize_cuda(cudaHandle);

        fprintf(stderr,"ssp_init_cuda: cusparse initialization failed.\n");
        return NULL;
    }

    status = cusparseCreateMatDescr(&cudaHandle->cusparse_matDescr); 
    if (status != CUSPARSE_STATUS_SUCCESS) {
        ssp_finalize_cuda(cudaHandle);

        fprintf(stderr,"ssp_init_cuda: cusparse matrix setup failed.\n");
        return NULL;
    }       
    cusparseSetMatType(cudaHandle->cusparse_matDescr,CUSPARSE_MATRIX_TYPE_GENERAL);
    cusparseSetMatIndexBase(cudaHandle->cusparse_matDescr,CUSPARSE_INDEX_BASE_ZERO);


    return cudaHandle;
}

// finalize CUDA
void ssp_finalize_cuda(ssp_cuda *cudaHandle) {
    if (!cudaHandle)
        return;

    if (cudaHandle->cusparse_handle)
        cusparseDestroy(cudaHandle->cusparse_handle);
    if (cudaHandle->cusparse_matDescr) 
        cusparseDestroyMatDescr(cudaHandle->cusparse_matDescr);

    free(cudaHandle);

    cudaDeviceReset();
}


include/ssparse.h
/*
 * This file is part of ssparse.
 *
 * Copyright (c) 2006, Timothy A. Davis.
 * File derived from:
 * CXSparse: a Concise Sparse matrix package - Extended.
 * http://www.suitesparse.com
 * 
 * Modified by Stefan Sommer (shso@elektro.dtu.dk), November 2012
 *
 * ssparse is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * ssparse is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this Module; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */
 
#ifndef _SSPS_H
#define _SSPS_H
#include <stdlib.h>
#include <limits.h>
#include <math.h>
#include <stdio.h>
#ifdef MATLAB_MEX_FILE
#include "mex.h"
#endif
#ifdef SSP_CUDA
#include <cuda_runtime.h>
#include "cusparse_v2.h"
#endif

#include "SuiteSparse_config.h"
#include "cs.h"

#ifdef __cplusplus
extern "C" {
#endif

#ifdef CS_LONG
#ifdef CS_COMPLEX
#define SSP_NAME(nm) ssp_cl ## nm
#define ssp ssp_cl
#else
#define SSP_NAME(nm) ssp_dl ## nm
#define ssp ssp_dl
#endif
#else
#ifdef CS_COMPLEX
#define SSP_NAME(nm) ssp_ci ## nm
#define ssp ssp_ci
#else
#define SSP_NAME(nm) ssp_di ## nm
#define ssp ssp_di
#endif
#endif

int ssp_di_splsolve (cs_di *L, const cs_di *B, int k, int *xi, double *x,
    const int *pinv) ;
cs_long_t ssp_dl_splsolve (cs_dl *L, const cs_dl *B, cs_long_t k, cs_long_t *xi,
    double *x, const cs_long_t *pinv) ;
int ssp_ci_splsolve (cs_ci *L, const cs_ci *B, int k, int *xi, 
    cs_complex_t *x, const int *pinv) ;
cs_long_t ssp_cl_splsolve (cs_cl *L, const cs_cl *B, cs_long_t k, cs_long_t *xi, 
    cs_complex_t *x, const cs_long_t *pinv) ;

int ssp_di_sputsolve (cs_di *L, const cs_di *B, int k, int *xi, double *x,
    const int *pinv) ;
cs_long_t ssp_dl_sputsolve (cs_dl *L, const cs_dl *B, cs_long_t k, cs_long_t *xi,
    double *x, const cs_long_t *pinv) ;
int ssp_ci_sputsolve (cs_ci *L, const cs_ci *B, int k, int *xi, 
    cs_complex_t *x, const int *pinv) ;
cs_long_t ssp_cl_sputsolve (cs_cl *L, const cs_cl *B, cs_long_t k, cs_long_t *xi, 
    cs_complex_t *x, const cs_long_t *pinv) ;

CS_ENTRY ssp_di_spdot (CS_INT *xi, CS_ENTRY *x, CS_ENTRY *y, CS_INT n) ;
CS_ENTRY ssp_dl_spdot (CS_INT *xi, CS_ENTRY *x, CS_ENTRY *y, CS_INT n) ;
CS_ENTRY ssp_ci_spdot (CS_INT *xi, CS_ENTRY *x, CS_ENTRY *y, CS_INT n) ;
CS_ENTRY ssp_cl_spdot (CS_INT *xi, CS_ENTRY *x, CS_ENTRY *y, CS_INT n) ;

#define ssp_splsolve SSP_NAME (_splsolve)
#define ssp_sputsolve SSP_NAME (_sputsolve)
#define ssp_spdot SSP_NAME (_spdot)

#ifdef SSP_CUDA
typedef struct {
    cusparseHandle_t cusparse_handle;
    cusparseMatDescr_t cusparse_matDescr;
} ssp_cuda;

ssp_cuda * ssp_init_cuda(); // initialize CUDA
void ssp_finalize_cuda(ssp_cuda *cudaHandle); // close CUDA

int ssp_spdot_cuda_init (int *xpHost, int *xiHost, CS_INT n, CS_INT m, CS_INT cols, int **xpDev, int **xiDev, cuDoubleComplex **xDev, cuDoubleComplex **yDev, cuDoubleComplex **resDev, ssp_cuda *cudaHandle);
void ssp_spdot_cuda_finalize (int **xpDev, int **xiDev, cuDoubleComplex **xDev, cuDoubleComplex **yDev, cuDoubleComplex **resDev);
int ssp_spdot_cuda (cuDoubleComplex *xHost, cuDoubleComplex *yHost, cuDoubleComplex *resHost, cuDoubleComplex alpha, cuDoubleComplex beta, int *xpDev, int *xiDev,
        cuDoubleComplex *xDev, cuDoubleComplex *yDev, cuDoubleComplex *resDev, CS_INT n, CS_INT m, CS_INT cols, CS_INT nnzx, ssp_cuda *cudaHandle);

#endif


#ifdef __cplusplus
}
#endif
#endif


src/ssp_splsolve.c
/*
 * This file is part of ssparse.
 *
 * Copyright (c) 2006, Timothy A. Davis.
 * File derived from:
 * CXSparse: a Concise Sparse matrix package - Extended.
 * http://www.suitesparse.com
 * 
 * Modified by Stefan Sommer (shso@elektro.dtu.dk), November 2012
 *
 * ssparse is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * ssparse is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this Module; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */
 
#include "ssparse.h"
/* solve Lx=b(:,k), where L is lower triangular with unit diagonal */
CS_INT ssp_splsolve (cs *G, const cs *B, CS_INT k, CS_INT *reach, CS_ENTRY *x, const CS_INT *pinv)
{
    CS_INT j, J, p, q, px, n, *Gp, *Gi, *Bp, *Bi ;
    CS_ENTRY *Gx, *Bx ;
    CS_INT top = reach[0];
    CS_INT *xi = reach+1;

    if (!CS_CSC (G) || !CS_CSC (B) || !xi || !x) return (-1) ;
    Gp = G->p ; Gi = G->i ; Gx = G->x ; n = G->n ;
    Bp = B->p ; Bi = B->i ; Bx = B->x ;
    // top = cs_reach (G, B, k, xi, pinv) ;        /* xi[top..n-1]=Reach(B(:,k)) */
    
    for (p = top ; p < n ; p++) x [xi [p]] = 0 ;    /* clear x */
    for (p = Bp [k] ; p < Bp [k+1] ; p++) x [Bi [p]] = Bx [p] ; /* scatter B */
    for (px = top ; px < n ; px++)
    {
        j = xi [px] ;                               /* x(j) is nonzero */
        J = pinv ? (pinv [j]) : j ;                 /* j maps to col J of G */
        if (J < 0) continue ;                       /* column J is empty */
        p = Gp [J]+1 ;            /* lo: L(j,j) 1st entry */
        q = Gp [J+1] ;
        for ( ; p < q ; p++)
        {
            x [Gi [p]] -= Gx [p] * x [j] ;          /* x(i) -= G(i,j) * x(j) */
        }
    }
    return (top) ;                                  /* return top of stack */
}


src/ssp_sputsolve.c
/*
 * This file is part of ssparse.
 *
 * Copyright (c) 2006, Timothy A. Davis.
 * File derived from:
 * CXSparse: a Concise Sparse matrix package - Extended.
 * http://www.suitesparse.com
 * 
 * Modified by Stefan Sommer (shso@elektro.dtu.dk), November 2012
 *
 * ssparse is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * ssparse is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this Module; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */
 
#include "ssparse.h"
/* solve Ux=b(:,k), where U is lower triangular with possible non-unit diagonal */
CS_INT ssp_sputsolve (cs *G, const cs *B, CS_INT k, CS_INT *reach, CS_ENTRY *x, const CS_INT *pinv)
{
    CS_INT j, J, p, q, px, n, *Gp, *Gi, *Bp, *Bi ;
    CS_ENTRY *Gx, *Bx ;
    CS_INT top = reach[0];
    CS_INT *xi = reach+1;

    if (!CS_CSC (G) || !CS_CSC (B) || !xi || !x) return (-1) ;
    Gp = G->p ; Gi = G->i ; Gx = G->x ; n = G->n ;
    Bp = B->p ; Bi = B->i ; Bx = B->x ;
    
    for (p = top ; p < n ; p++) x [xi [p]] = 0 ;    /* clear x */
    for (p = Bp [k] ; p < Bp [k+1] ; p++) x [Bi [p]] = Bx [p] ; /* scatter B */
#if 0
    for (px = top ; px < n ; px++)
    {
        j = xi [px] ;                               /* x(j) is nonzero */
        J = pinv ? (pinv [j]) : j ;                 /* j maps to col J of G */
        if (J < 0) continue ;                       /* column J is empty */
        p = Gp [J] ;
        q = Gp [J+1]-1 ;
        for ( ; p < q ; p++)
        {
            //if (x [Gi [p]] != 0)
                x [j] -= CS_CONJ(Gx [p]) * x [Gi [p]] ;          /* x(j) -= G(i,j) * x(i) */
        }
        x [j] /= CS_CONJ(Gx [Gp [J+1]-1]) ;
    }
    return (top) ;                                  /* return top of stack */
#else
    CS_INT lo = 1;
    for (px = top ; px < n ; px++)
    {
        j = xi [px] ;                               /* x(j) is nonzero */
        J = pinv ? (pinv [j]) : j ;                 /* j maps to col J of G */
        if (J < 0) continue ;                       /* column J is empty */
        x [j] /= Gx [lo ? (Gp [J]) : (Gp [J+1]-1)] ;/* x(j) /= G(j,j) */
        p = lo ? (Gp [J]+1) : (Gp [J]) ;            /* lo: L(j,j) 1st entry */
        q = lo ? (Gp [J+1]) : (Gp [J+1]-1) ;        /* up: U(j,j) last entry */
        for ( ; p < q ; p++)
        {
            x [Gi [p]] -= Gx [p] * x [j] ;          /* x(i) -= G(i,j) * x(j) */
        }
    }
    return (top) ;                                  /* return top of stack */
#endif
}


src/ssp_spdot_cuda.cpp
/*
 * This file is part of ssparse.
 *
 * Copyright (C) 2012, Technical University of Denmark
 * https://github.com/nefan/ssparse
 *
 * ssparse is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * ssparse is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this Module; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "ssparse.h"

/* dot product between sets of dense vectors x,y with x having non-zero pattern xi 
   CUDA accelerated version */
int ssp_spdot_cuda (cuDoubleComplex *xHost, cuDoubleComplex *yHost, cuDoubleComplex *resHost, 
        cuDoubleComplex alpha, cuDoubleComplex beta, int *xpDev, int *xiDev,
        cuDoubleComplex *xDev, cuDoubleComplex *yDev, cuDoubleComplex *resDev, 
        CS_INT n, CS_INT m, CS_INT cols, CS_INT nnzx, ssp_cuda *cudaHandle)
{
    cudaError_t stat1 = cudaMemcpy(xDev, xHost, (size_t)nnzx*sizeof(cuDoubleComplex), cudaMemcpyHostToDevice);
    cudaError_t stat2 = cudaMemcpy(yDev, yHost, (size_t)n*cols*sizeof(cuDoubleComplex), cudaMemcpyHostToDevice);
    if ((stat1 != cudaSuccess) || (stat2 != cudaSuccess)) {
        fprintf(stderr,"ssp_spdot_cuda: host->device memory copy failed.\n");
        return false;
    }

    if (beta.x != 0 && beta.y != 0) {
        stat1 = cudaMemcpy(resDev, resHost, (size_t)std::max(n,m)*cols*sizeof(cuDoubleComplex), cudaMemcpyHostToDevice);
        if ((stat1 != cudaSuccess)) {
            fprintf(stderr,"ssp_spdot_cuda: host->device memory copy failed.\n");
            return false;
        }
    }

    // do inner product computation
    cusparseStatus_t status = cusparseZcsrmm(cudaHandle->cusparse_handle, 
            CUSPARSE_OPERATION_NON_TRANSPOSE, 
            m, cols, n, nnzx, &alpha, 
            cudaHandle->cusparse_matDescr, 
            xDev, xpDev, xiDev, yDev, 
            n, &beta, resDev, std::max(n,m));
    if (status != CUSPARSE_STATUS_SUCCESS) {
        fprintf(stderr,"ssp_spdot_cuda: matrix-matrix multiplication failed.\n");
        return false;
    }  
    
    stat1 = cudaMemcpy(resHost, resDev, (size_t)std::max(n,m)*cols*sizeof(cuDoubleComplex), cudaMemcpyDeviceToHost);
    if ((stat1 != cudaSuccess)) {
        fprintf(stderr,"ssp_spdot_cuda: device->host memory copy failed. %d\n",stat1);
        return false;
    }

    return true;
}

// allocate memory etc.
int ssp_spdot_cuda_init (int *xpHost, int *xiHost, CS_INT n, CS_INT m, CS_INT cols, 
        int **xpDev, int **xiDev, 
        cuDoubleComplex **xDev, cuDoubleComplex **yDev, cuDoubleComplex **resDev, 
        ssp_cuda *cudaHandle) {

    int nnzx = xpHost[m];

    cudaError_t stat1 = cudaMalloc(xpDev, (m+1)*sizeof(int));
    cudaError_t stat2 = cudaMalloc(xiDev, nnzx*sizeof(int));
    cudaError_t stat3 = cudaMalloc(xDev, nnzx*sizeof(cuDoubleComplex));
    cudaError_t stat4 = cudaMalloc(yDev, n*cols*sizeof(cuDoubleComplex));
    cudaError_t stat5 = cudaMalloc(resDev, std::max(n,m)*cols*sizeof(cuDoubleComplex));

    if ((stat1 != cudaSuccess) || (stat2 != cudaSuccess) ||
            (stat3 != cudaSuccess) || (stat4 != cudaSuccess) ||
            (stat5 != cudaSuccess)) {

        ssp_spdot_cuda_finalize(xpDev,xiDev,xDev,yDev,resDev);

        fprintf(stderr,"ssp_spdot_cuda_init: memory allocation failed.\n");
        return false;
    }

    stat1 = cudaMemcpy(*xpDev, xpHost, (size_t)(m+1)*sizeof(int), cudaMemcpyHostToDevice);
    stat2 = cudaMemcpy(*xiDev, xiHost, (size_t)nnzx*sizeof(int), cudaMemcpyHostToDevice);
    if ((stat1 != cudaSuccess) || (stat2 != cudaSuccess)) {

        ssp_spdot_cuda_finalize(xpDev,xiDev,xDev,yDev,resDev);

        fprintf(stderr,"ssp_spdot_cuda_init: host->device memory copy failed.\n");
        return false;
    }

    return true;
}

void ssp_spdot_cuda_finalize (int **xpDev, int **xiDev, cuDoubleComplex **xDev, cuDoubleComplex **yDev, cuDoubleComplex **resDev) {
    if (*xpDev) cudaFree(*xpDev); *xpDev = 0;
    if (*xiDev) cudaFree(*xiDev); *xiDev = 0;
    if (*xDev) cudaFree(*xDev); *xDev = 0;
    if (*yDev) cudaFree(*yDev); *yDev = 0;
    if (*resDev) cudaFree(*resDev); *resDev = 0;
}
